### DISCUSSION TITLE
Cannot find reference 'as_array'/'get_buffer_as_array' in 'OpusFile | OpusFile'

### DISCUSSION THREAD

**Discussion by michel538** [TeamPyOgg/PyOgg#67] [Q&A]
Hello,

I would like to run 04-opus-file-stream.py test as to learn pyogg.

But I meet with following error:

Cannot find reference 'as_array'/'get_buffer_as_array' in 'OpusFile | OpusFile'

here:

    array = opus_file.as_array()

and:

    buf = stream.get_buffer_as_array()

Question are these still supported in the pyogg?

I hope this is the good forum to ask beginners question.

Thanks

**Comment by mattgwwalker** [comment]
Hi Michel,

Sorry I took so long to reply, I didn't see your question until just now.

I am not certain, but I suspect that the issue lies in the version of PyOgg you've installed.  I imagine you've run "pip install pyogg", which installed version 0.6.14a1.  What happens if you install the version in the GitHub repository?

    git clone https://github.com/TeamPyOgg/PyOgg.git
    pip install -e PyOgg

That should allow you to run `04-opus-file-stream.py` without issues.

Please report back to tell me how you go.

Cheers,

Matthew

**Reply by michel538** [reply]
Hello Matthew,

I  would like to thank you for you help - your suggestion solved the issue.

I would appreciate if you could provide me with more guidance .

The application is to stream ogg/opus audios to a inhouse ogg decoder **in real time**.

The logic I consider would be:

- to breakdown ogg audios in chunk
- to evaluate the duration of each chunk 
- to wait "the duration" at each streaming of chunk

Do you know if pyogg.py can provide means to do step 1 and 2?

Do you mind explaining how I should understand the following: 

Read 330360 samples from the OpusFileStream
(in 174 reads averaging 0.10 milliseconds each).
OpusFileStream data was identical to OpusFile data,
however there was more data in the OpusFile than
in the OpusFileStream.

Again I hope the forum is right fr my questions.

Thanks

Michel

**Comment by mattgwwalker** [comment]
PyOgg can be used to achieve real-time encoding and/or decoding of Opus audio.  Indeed, that's what I use it for personally.  Most of these features are not yet released however.  An official release, including all the recent Opus-related work, will happen soon. 

I would also expect that there would be no problem at all with real time encoding and decoding of OggOpus audio, especially if you're using the master branch from the git repository.

However, I'm not entirely sure what you're trying to do.  Are you trying effectively to write a Python program to play OggOpus files from, say, a network source?  Are there synchronization concerns for playback? 

I would recommend that you consider [OpusFileStream](https://pyogg.readthedocs.io/en/latest/api/opus.html#opusfilestream).

`OpusFileStream.get_buffer()` or `OpusFileStream.get_buffer_as_array()` will get you the next "chunk" of PCM (decoded audio).

The duration of that PCM can be easily calculated.  First find the length of the chuck in samples, then divide it by 48,000 (which is the number of samples per second for Opus-encoded audio).  That gives you the length in seconds.  Something like the following:

    array = stream.get_buffer_as_array()
    length_samples = len(array)
    samples_per_second = stream.frequency
    length_seconds = length_samples / samples_per_second
    
If your target is playback, then I'd recommend you consider [simpleaudio](https://simpleaudio.readthedocs.io/en/latest/).  That will allow you to play back in real time.

Finally, the last point in your question is concerning to me:
"however there was more data in the OpusFile than in the OpusFileStream".  That's an error condition if the two files you're comparing are identical.  Can you tell me more about how you got to there?

I hope that's helpful.

Matthew

**Comment by mattgwwalker** [comment]
I didn't answer your last question.  How should you understand

> Read 330360 samples from the OpusFileStream
(in 174 reads averaging 0.10 milliseconds each).
OpusFileStream data was identical to OpusFile data,
however there was more data in the OpusFile than
in the OpusFileStream.

OpusFile reads the entire file into memory.

OpusFileStream reads "chunks" of the file at a time until it's read the whole file.  In the case of your text, it's read 174 chunks and each chunk took on average 0.1 milliseconds to read.

OpusFileStream read a total of 330360 samples.  However OpusFile read more samples.

Finally, the first 330360 samples read by both OpusFile and OpusFileStream were identical.

**Comment by mattgwwalker** [comment]
If you have other questions would you mind creating a new discussion for each topic?  I think that would help others who were searching these archives.

Cheers,

Matthew
